numpy/core/src/multiarray/ctors.c

```diff
@@ -2853,12 +2847,32 @@ PyArray_Zeros(int nd, npy_intp *dims, PyArray_Descr *type, int is_f_order)
     if (!type) {
         type = PyArray_DescrFromType(NPY_DEFAULT_TYPE);
     }
+    /* Resize flexible types like string and unicode to hold one character */
+    if (type->elsize == -1) {
+        /* manually invoke np.dtype((type, 1)) */
+        PyObject *tuple = PyTuple_New(2);
+        PyObject *one = PyInt_FromLong(1);
+        int status;
+
+        if (tuple == NULL || one == NULL) {
+            Py_DECREF(type);
+            return NULL;
+        }
+        PyTuple_SET_ITEM(tuple, 0, type);
+        PyTuple_SET_ITEM(tuple, 1, one);
+
+        status = PyArray_DescrConverter(tuple, &type);
```

Comment:
yes but PyArray_DescrConverter would still be working inplace, its output type is the same object as its input type. It probably works fine but that should be checked.